// @skyscanner/backpack-web/bpk-component-card/src/BpkDividedCard.js
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { cssModules } from "../../bpk-react-utils";
import BpkCard from "./BpkCard";
import { CardContext } from "./CardContext";
import STYLES from "./BpkDividedCard.module.css";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
const getClassName = cssModules(STYLES);
export const ORIENTATION = {
  horizontal: 'horizontal',
  vertical: 'vertical'
};
const BpkDividedCard = ({
  href = null,
  isElevated = true,
  orientation = ORIENTATION.horizontal,
  primaryContent,
  secondaryContent,
  ...rest
}) => {
  const isVertical = orientation === ORIENTATION.vertical;
  const containerClassName = getClassName(isVertical ? 'bpk-divided-card--vertical-container' : 'bpk-divided-card--horizontal-container');
  const contentClassName = getClassName('bpk-divided-card--content', isVertical && 'bpk-divided-card--vertical-content');
  const secondaryContentClassName = getClassName(isVertical ? 'bpk-divided-card__secondary--vertical' : 'bpk-divided-card__secondary--horizontal');
  return /*#__PURE__*/_jsx(CardContext.Provider, {
    value: {
      elevated: isElevated
    },
    children: /*#__PURE__*/_jsx("div", {
      className: containerClassName,
      children: /*#__PURE__*/_jsx(BpkCard, {
        href: href,
        padded: false,
        ...rest,
        children: /*#__PURE__*/_jsxs("div", {
          className: contentClassName,
          children: [/*#__PURE__*/_jsx("div", {
            className: getClassName(!isVertical && 'bpk-divided-card__primary--horizontal'),
            children: primaryContent
          }), /*#__PURE__*/_jsx("div", {
            className: secondaryContentClassName,
            children: secondaryContent
          })]
        })
      })
    })
  });
};
export default BpkDividedCard;
