// @skyscanner/backpack-web/bpk-component-modal/src/BpkModal.js
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Portal, cssModules, isDeviceIphone } from "../../bpk-react-utils";
import { withScrim } from "../../bpk-scrim-utils";
import BpkModalInner, { MODAL_STYLING } from "./BpkModalInner";
import STYLES from "./BpkModal.module.css";
import { jsx as _jsx } from "react/jsx-runtime";
const getClassName = cssModules(STYLES);
const ScrimBpkModalInner = withScrim(BpkModalInner);
const BpkModal = ({
  accessoryView = null,
  className = null,
  closeLabel = '',
  closeOnEscPressed = true,
  closeOnScrimClick = true,
  closeText = null,
  contentClassName = null,
  dialogRef = () => null,
  fullScreen = false,
  fullScreenOnMobile = true,
  isIphone = isDeviceIphone(),
  isOpen,
  modalStyle = MODAL_STYLING.default,
  onClose = () => null,
  padded = true,
  renderTarget = null,
  showHeader = true,
  title = null,
  wide = false,
  ...rest
}) => {
  const containerClass = [getClassName('bpk-modal__container')];
  if (fullScreen || isIphone) {
    containerClass.push(getClassName('bpk-modal__container--full-screen'));
  } else if (fullScreenOnMobile) {
    containerClass.push(getClassName('bpk-modal__container--full-screen-mobile'));
  }
  return /*#__PURE__*/_jsx(Portal, {
    isOpen: isOpen,
    onClose: onClose,
    renderTarget: renderTarget,
    closeOnEscPressed: closeOnEscPressed,
    children: /*#__PURE__*/_jsx(ScrimBpkModalInner, {
      onClose: onClose,
      fullScreenOnMobile: fullScreenOnMobile,
      fullScreen: fullScreen,
      closeOnScrimClick: closeOnScrimClick,
      containerClassName: containerClass.join(' '),
      isIphone: isIphone,
      title: title,
      className: className,
      contentClassName: contentClassName,
      closeLabel: closeLabel,
      closeText: closeText,
      wide: wide,
      showHeader: showHeader,
      padded: padded,
      accessoryView: accessoryView,
      dialogRef: dialogRef,
      modalStyle: modalStyle,
      ...rest
    })
  });
};
export default BpkModal;
