// @skyscanner/backpack-web/bpk-component-list/src/BpkList.js
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */import PropTypes from 'prop-types';
import { cssModules } from "../../bpk-react-utils";
import STYLES from "./BpkList.module.css";
import { jsx as _jsx } from "react/jsx-runtime";
const getClassName = cssModules(STYLES);
const BpkList = props => {
  const {
    ariaLabel,
    ariaLabelledby,
    children,
    className,
    ordered,
    title
  } = props;
  const ListElements = ordered ? 'ol' : 'ul';
  const classNames = getClassName('bpk-list', className);
  return /*#__PURE__*/_jsx(ListElements, {
    "aria-label": ariaLabel,
    "aria-labelledby": ariaLabelledby,
    title: title,
    className: classNames,
    children: children
  });
};
BpkList.propTypes = {
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,
  ordered: PropTypes.bool,
  className: PropTypes.string,
  ariaLabel: PropTypes.string,
  ariaLabelledby: PropTypes.string,
  title: PropTypes.string
};
BpkList.defaultProps = {
  ordered: false,
  className: null,
  ariaLabel: null,
  ariaLabelledby: null,
  title: null
};
export default BpkList;
