// @skyscanner/backpack-web/bpk-component-section-list/src/BpkSectionListItem.js
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */import PropTypes from 'prop-types';
import { withRtlSupport } from "../../bpk-component-icon";
import BpkLargeChevronRightIcon from "../../bpk-component-icon/lg/chevron-right";
import { cssModules } from "../../bpk-react-utils";
import STYLES from "./BpkSectionListItem.module.css";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
const BpkLargeChevronRightIconWithRtlSupport = withRtlSupport(BpkLargeChevronRightIcon);
const getClassName = cssModules(STYLES);
const BpkSectionListItem = props => {
  const {
    blank,
    children,
    className,
    href,
    onClick,
    ...rest
  } = props;
  const classNames = [getClassName('bpk-section-list-item', (href || onClick) && 'bpk-section-list-item--interactive', className)];
  if (href) {
    const target = blank ? '_blank' : null;
    return (
      /*#__PURE__*/
      // $FlowFixMe[cannot-spread-inexact] - inexact rest. See 'decisions/flowfixme.md'.
      _jsxs("a", {
        href: href,
        target: target,
        onClick: onClick,
        className: classNames.join(' '),
        ...rest,
        children: [children, /*#__PURE__*/_jsx("span", {
          className: getClassName('bpk-section-list-item__chevron'),
          children: /*#__PURE__*/_jsx(BpkLargeChevronRightIconWithRtlSupport, {})
        })]
      })
    );
  }
  if (onClick) {
    return (
      /*#__PURE__*/
      // $FlowFixMe[cannot-spread-inexact] - inexact rest. See 'decisions/flowfixme.md'.
      _jsxs("button", {
        type: "button",
        onClick: onClick,
        className: classNames.join(' '),
        ...rest,
        children: [children, /*#__PURE__*/_jsx("span", {
          className: getClassName('bpk-section-list-item__chevron'),
          children: /*#__PURE__*/_jsx(BpkLargeChevronRightIconWithRtlSupport, {})
        })]
      })
    );
  }
  return (
    /*#__PURE__*/
    // $FlowFixMe[cannot-spread-inexact] - inexact rest. See 'decisions/flowfixme.md'.
    _jsx("div", {
      className: classNames.join(' '),
      ...rest,
      children: children
    })
  );
};
BpkSectionListItem.propTypes = {
  children: PropTypes.node.isRequired,
  blank: PropTypes.bool,
  className: PropTypes.string,
  href: PropTypes.string,
  onClick: PropTypes.func
};
BpkSectionListItem.defaultProps = {
  blank: false,
  className: null,
  href: null,
  onClick: null
};
export default BpkSectionListItem;
