// @skyscanner/backpack-web/bpk-component-calendar/src/BpkCalendarGridHeader.js
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { PureComponent } from 'react';
import { cssModules } from "../../bpk-react-utils";
import { orderDaysOfWeek } from "./date-utils";
import STYLES from "./BpkCalendarGridHeader.module.css";
import { jsx as _jsx } from "react/jsx-runtime";
const getClassName = cssModules(STYLES);
/*
  WeekDay - table header cells such as "Mon", "Tue", "Wed"...
*/
const WeekDayComponent = ({
  Element,
  weekDay,
  weekDayKey = 'nameAbbr'
}) => /*#__PURE__*/_jsx(Element
// TODO: className to be removed
// eslint-disable-next-line @skyscanner/rules/forbid-component-props
, {
  className: getClassName('bpk-calendar-header__weekday'),
  title: weekDay.name,
  children: /*#__PURE__*/_jsx("span", {
    children: weekDay[weekDayKey]
  })
});
class BpkCalendarGridHeader extends PureComponent {
  static defaultProps = {
    className: null,
    weekDayKey: 'nameAbbr'
  };
  render() {
    const {
      className,
      weekDayKey,
      weekStartsOn
    } = this.props;
    const daysOfWeek = orderDaysOfWeek(this.props.daysOfWeek, weekStartsOn);
    const classNames = [getClassName('bpk-calendar-header')];
    if (className) {
      classNames.push(className);
    }
    return /*#__PURE__*/_jsx("header", {
      className: classNames.join(' '),
      "aria-hidden": true,
      children: /*#__PURE__*/_jsx("ol", {
        className: getClassName('bpk-calendar-header__week'),
        children: daysOfWeek.map(weekDay => /*#__PURE__*/_jsx(WeekDayComponent, {
          Element: "li",
          weekDay: weekDay,
          weekDayKey: weekDayKey
        }, weekDay.index))
      })
    });
  }
}
export default BpkCalendarGridHeader;
