// @skyscanner/backpack-web/bpk-component-spinner/src/BpkSpinner.js
import React from "react";
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import PropTypes from 'prop-types';
var SmSpinner = function SmSpinner(props) {
  return /*#__PURE__*/_jsxs("svg", {
    ...props,
    children: [/*#__PURE__*/_jsx("path", {
      d: "M3 7H1a1 1 0 0 0 0 2h2a1 1 0 0 0 0-2Z",
      opacity: ".45"
    }), /*#__PURE__*/_jsx("path", {
      d: "M15 7h-2a1 1 0 1 0 0 2h2a1 1 0 1 0 0-2Z"
    }), /*#__PURE__*/_jsx("path", {
      d: "M7 13v2a1 1 0 1 0 2 0v-2a1 1 0 1 0-2 0Z",
      opacity: ".3"
    }), /*#__PURE__*/_jsx("path", {
      d: "M7 1v2a1 1 0 0 0 2 0V1a1 1 0 0 0-2 0Z",
      opacity: ".65"
    }), /*#__PURE__*/_jsx("path", {
      d: "m9.634 12.83 1 1.732a1 1 0 1 0 1.732-1l-1-1.732a1 1 0 0 0-1.732 1Z",
      opacity: ".25"
    }), /*#__PURE__*/_jsx("path", {
      d: "m3.634 2.438 1 1.732a1 1 0 1 0 1.732-1l-1-1.732a1 1 0 1 0-1.732 1Z",
      opacity: ".55"
    }), /*#__PURE__*/_jsx("path", {
      d: "m11.83 11.366 1.732 1a1 1 0 1 0 1-1.732l-1.732-1a1 1 0 1 0-1 1.732Z",
      opacity: ".2"
    }), /*#__PURE__*/_jsx("path", {
      d: "m1.438 5.366 1.732 1a1 1 0 1 0 1-1.732l-1.732-1a1 1 0 1 0-1 1.732Z",
      opacity: ".5"
    }), /*#__PURE__*/_jsx("path", {
      d: "m12.83 6.366 1.732-1a1 1 0 0 0-1-1.732l-1.732 1a1 1 0 0 0 1 1.732Z",
      opacity: ".85"
    }), /*#__PURE__*/_jsx("path", {
      d: "m2.438 12.366 1.732-1a1 1 0 1 0-1-1.732l-1.732 1a1 1 0 1 0 1 1.732Z",
      opacity: ".4"
    }), /*#__PURE__*/_jsx("path", {
      d: "m11.366 4.17 1-1.732a1 1 0 0 0-1.732-1l-1 1.732a1 1 0 0 0 1.732 1Z",
      opacity: ".75"
    }), /*#__PURE__*/_jsx("path", {
      d: "m5.366 14.562 1-1.732a1 1 0 1 0-1.732-1l-1 1.732a1 1 0 0 0 1.732 1Z",
      opacity: ".35"
    })]
  });
};
SmSpinner.defaultProps = {
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 16 16"
};
import { cssModules } from "../../bpk-react-utils";
import SPINNER_TYPES from "./spinnerTypes";
import STYLES from "./BpkSpinner.module.css";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
const getClassName = cssModules(STYLES);
const BpkSpinner = props => {
  const {
    alignToButton,
    className,
    type,
    ...rest
  } = props;
  const classNames = getClassName('bpk-spinner', `bpk-spinner--${type}`, alignToButton && 'bpk-spinner--align-to-button', className);
  return /*#__PURE__*/_jsx("span", {
    className: classNames,
    children: /*#__PURE__*/_jsx(SmSpinner, {
      ...rest
    })
  });
};
BpkSpinner.propTypes = {
  type: PropTypes.oneOf(Object.keys(SPINNER_TYPES)),
  className: PropTypes.string,
  alignToButton: PropTypes.bool
};
BpkSpinner.defaultProps = {
  type: SPINNER_TYPES.dark,
  className: null,
  alignToButton: false
};
export default BpkSpinner;
