// @skyscanner/backpack-web/bpk-component-button/src/BpkButton.js
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import PropTypes from 'prop-types';
import BpkButtonBase, { BUTTON_TYPES } from "./BpkButtonBase";
import { propTypes, defaultProps } from "./common-types";
import { jsx as _jsx } from "react/jsx-runtime";
const BpkButton = props => {
  const {
    destructive,
    featured,
    link,
    linkOnDark,
    primaryOnDark,
    primaryOnLight,
    secondary,
    secondaryOnDark,
    ...rest
  } = props;
  if (primaryOnDark) {
    return /*#__PURE__*/_jsx(BpkButtonBase, {
      type: BUTTON_TYPES.primaryOnDark,
      ...rest
    });
  }
  if (primaryOnLight) {
    return /*#__PURE__*/_jsx(BpkButtonBase, {
      type: BUTTON_TYPES.primaryOnLight,
      ...rest
    });
  }
  if (secondary) {
    return /*#__PURE__*/_jsx(BpkButtonBase, {
      type: BUTTON_TYPES.secondary,
      ...rest
    });
  }
  if (secondaryOnDark) {
    return /*#__PURE__*/_jsx(BpkButtonBase, {
      type: BUTTON_TYPES.secondaryOnDark,
      ...rest
    });
  }
  if (destructive) {
    return /*#__PURE__*/_jsx(BpkButtonBase, {
      type: BUTTON_TYPES.destructive,
      ...rest
    });
  }
  if (featured) {
    return /*#__PURE__*/_jsx(BpkButtonBase, {
      type: BUTTON_TYPES.featured,
      ...rest
    });
  }
  if (link) {
    return /*#__PURE__*/_jsx(BpkButtonBase, {
      type: BUTTON_TYPES.link,
      ...rest
    });
  }
  if (linkOnDark) {
    return /*#__PURE__*/_jsx(BpkButtonBase, {
      type: BUTTON_TYPES.linkOnDark,
      ...rest
    });
  }
  return /*#__PURE__*/_jsx(BpkButtonBase, {
    ...rest
  });
};
BpkButton.defaultProps = {
  ...defaultProps,
  primaryOnDark: false,
  primaryOnLight: false,
  secondary: false,
  secondaryOnDark: false,
  destructive: false,
  featured: false,
  link: false,
  linkOnDark: false
};
BpkButton.propTypes = {
  ...propTypes,
  primaryOnDark: PropTypes.bool,
  primaryOnLight: PropTypes.bool,
  secondary: PropTypes.bool,
  secondaryOnDark: PropTypes.bool,
  destructive: PropTypes.bool,
  featured: PropTypes.bool,
  link: PropTypes.bool,
  linkOnDark: PropTypes.bool
};
export default BpkButton;
