// @skyscanner/backpack-web/bpk-component-ticket/src/BpkTicket.js
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */import PropTypes from 'prop-types';
import { cssModules } from "../../bpk-react-utils";
import STYLES from "./BpkTicket.module.css";
import { jsx as _jsx } from "react/jsx-runtime";
const getClassName = cssModules(STYLES);
const BpkTicket = props => {
  const {
    children,
    className,
    href,
    padded,
    stub,
    stubClassName,
    stubProps,
    vertical,
    ...rest
  } = props;
  const classNames = getClassName('bpk-ticket', className, vertical && 'bpk-ticket--vertical');
  const mainClassNames = getClassName('bpk-ticket__paper', 'bpk-ticket__main', padded && 'bpk-ticket__main--padded', vertical && 'bpk-ticket__main--vertical', !vertical && 'bpk-ticket__main--horizontal');
  const mainInnerClassNames = getClassName('bpk-ticket__main-inner', vertical && 'bpk-ticket__main-inner--vertical', !vertical && 'bpk-ticket__main-inner--horizontal');
  const stubClassNames = getClassName('bpk-ticket__paper', 'bpk-ticket__stub', stubClassName, padded && 'bpk-ticket__stub--padded', vertical && 'bpk-ticket__stub--vertical', !vertical && 'bpk-ticket__stub--horizontal');
  const stubInnerClassNames = getClassName('bpk-ticket__stub-inner', vertical && 'bpk-ticket__stub-inner--vertical', !vertical && 'bpk-ticket__stub-inner--horizontal');
  const mainContent = padded ? children : /*#__PURE__*/_jsx("div", {
    className: mainInnerClassNames,
    children: children
  });
  const stubContent = padded ? stub : /*#__PURE__*/_jsx("div", {
    className: stubInnerClassNames,
    children: stub
  });
  const contents = [/*#__PURE__*/_jsx("div", {
    className: mainClassNames,
    children: mainContent
  }, "main"),
  /*#__PURE__*/
  // $FlowFixMe[cannot-spread-indexer] - inexact rest. See 'decisions/flowfixme.md'.
  _jsx("div", {
    className: stubClassNames,
    ...stubProps,
    children: stubContent
  }, "stub")];
  if (href) {
    return (
      /*#__PURE__*/
      // $FlowFixMe[cannot-spread-inexact] - inexact rest. See 'decisions/flowfixme.md'.
      _jsx("a", {
        href: href,
        className: classNames,
        ...rest,
        children: contents
      })
    );
  }
  return (
    /*#__PURE__*/
    // $FlowFixMe[cannot-spread-inexact] - inexact rest. See 'decisions/flowfixme.md'.
    _jsx("div", {
      role: "button",
      className: classNames,
      ...rest,
      children: contents
    })
  );
};
BpkTicket.propTypes = {
  children: PropTypes.node.isRequired,
  stub: PropTypes.node.isRequired,
  className: PropTypes.string,
  href: PropTypes.string,
  padded: PropTypes.bool,
  vertical: PropTypes.bool,
  stubClassName: PropTypes.string,
  stubProps: PropTypes.object // eslint-disable-line react/forbid-prop-types
};
BpkTicket.defaultProps = {
  className: null,
  href: null,
  padded: true,
  vertical: false,
  stubClassName: null,
  stubProps: {}
};
export default BpkTicket;
