// @skyscanner/backpack-web/bpk-component-map/src/BpkPriceMarker.js
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */import PropTypes from 'prop-types';
import BpkText, { TEXT_STYLES } from "../../bpk-component-text/src/BpkText";
import { cssModules } from "../../bpk-react-utils";
import BpkBasicMapMarker from "./BpkBasicMapMarker";
import { LatLongPropType } from "./common-types";
import STYLES from "./BpkPriceMarker.module.css";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
const getClassName = cssModules(STYLES);
export const PRICE_MARKER_STATUSES = {
  default: 'default',
  focused: 'focused',
  viewed: 'viewed'
};
const BpkPriceMarker = props => {
  const {
    arrowClassName,
    buttonProps,
    className,
    disabled,
    label,
    onClick,
    position,
    status,
    ...rest
  } = props;
  const markerWrapperClassNames = getClassName('bpk-price-marker__wrapper');
  const classNames = getClassName('bpk-price-marker', `bpk-price-marker-${status}`, onClick && 'bpk-price-marker--dynamic', className, disabled && `bpk-price-marker-disabled`);
  const arrowClassNames = getClassName('bpk-price-marker__arrow', `bpk-price-marker__arrow-${status}`, arrowClassName, disabled && `bpk-price-marker__arrow-disabled`);
  return (
    /*#__PURE__*/
    // $FlowFixMe[cannot-spread-inexact] - inexact rest. See 'decisions/flowfixme.md'.
    _jsx(BpkBasicMapMarker, {
      position: position,
      ...rest,
      children: /*#__PURE__*/_jsxs("button", {
        type: "button",
        className: markerWrapperClassNames,
        onClick: onClick,
        disabled: disabled,
        ...buttonProps,
        children: [/*#__PURE__*/_jsx("span", {
          className: classNames,
          children: /*#__PURE__*/_jsx(BpkText, {
            textStyle: status === PRICE_MARKER_STATUSES.focused ? TEXT_STYLES.label1 : TEXT_STYLES.label2,
            children: label
          })
        }), /*#__PURE__*/_jsx("div", {
          className: arrowClassNames
        })]
      })
    })
  );
};
BpkPriceMarker.propTypes = {
  label: PropTypes.string.isRequired,
  position: LatLongPropType.isRequired,
  arrowClassName: PropTypes.string,
  className: PropTypes.string,
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
  status: PropTypes.oneOf(Object.keys(PRICE_MARKER_STATUSES)),
  buttonProps: PropTypes.object // eslint-disable-line react/forbid-prop-types
};
BpkPriceMarker.defaultProps = {
  className: null,
  arrowClassName: null,
  onClick: null,
  disabled: false,
  status: PRICE_MARKER_STATUSES.default,
  buttonProps: null
};
export default BpkPriceMarker;
