// @skyscanner/backpack-web/bpk-component-label/src/BpkLabel.js
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */import PropTypes from 'prop-types';
import { cssModules } from "../../bpk-react-utils";
import STYLES from "./BpkLabel.module.css";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
const getClassName = cssModules(STYLES);
const BpkLabel = props => {
  const {
    children,
    className,
    disabled,
    required,
    valid,
    white,
    ...rest
  } = props;
  const invalid = valid === false;
  const classNames = getClassName('bpk-label', white && 'bpk-label--white', invalid && 'bpk-label--invalid', disabled && 'bpk-label--disabled', white && disabled && 'bpk-label--disabled--white', className);
  return (
    /*#__PURE__*/
    // $FlowFixMe[cannot-spread-inexact] - inexact rest. See 'decisions/flowfixme.md'.
    _jsxs("label", {
      className: classNames,
      ...rest,
      children: [children, !disabled && required && /*#__PURE__*/_jsx("span", {
        className: getClassName('bpk-label__asterisk'),
        children: "*"
      })]
    })
  );
};
BpkLabel.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  valid: PropTypes.bool,
  required: PropTypes.bool,
  white: PropTypes.bool
};
BpkLabel.defaultProps = {
  className: null,
  disabled: false,
  valid: null,
  required: false,
  white: false
};
export default BpkLabel;
