// @skyscanner/backpack-web/bpk-component-autosuggest/src/BpkAutosuggestSuggestion.js
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */import PropTypes from 'prop-types';
import { cssModules } from "../../bpk-react-utils";
import STYLES from "./BpkAutosuggest.module.css";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
const getClassName = cssModules(STYLES);
const BpkAutosuggestSuggestion = props => {
  const classNames = [getClassName('bpk-autosuggest__suggestion')];
  const {
    className,
    icon,
    indent,
    subHeading,
    tertiaryLabel,
    value,
    ...rest
  } = props;
  const Icon = icon;
  if (indent) {
    classNames.push(getClassName('bpk-autosuggest__suggestion--indent'));
  }
  if (className) {
    classNames.push(className);
  }
  return (
    /*#__PURE__*/
    // $FlowFixMe[cannot-spread-inexact] - inexact rest. See 'decisions/flowfixme.md'.
    _jsxs("section", {
      className: classNames.join(' '),
      ...rest,
      children: [icon ? /*#__PURE__*/_jsx("div", {
        className: getClassName('bpk-autosuggest__suggestion-icon'),
        children: /*#__PURE__*/_jsx(Icon, {})
      }) : null, /*#__PURE__*/_jsx("div", {
        className: getClassName('bpk-autosuggest__suggestion-content'),
        children: /*#__PURE__*/_jsxs("div", {
          className: getClassName('bpk-autosuggest__suggestion-inner'),
          children: [/*#__PURE__*/_jsxs("div", {
            children: [/*#__PURE__*/_jsx("span", {
              className: getClassName('bpk-autosuggest__suggestion-value'),
              children: value
            }), subHeading || tertiaryLabel ? /*#__PURE__*/_jsx("small", {
              className: getClassName('bpk-autosuggest__suggestion-sub-heading'),
              children: subHeading
            }) : null]
          }), tertiaryLabel ? /*#__PURE__*/_jsx("aside", {
            className: getClassName('bpk-autosuggest__suggestion-tertiary-label'),
            children: tertiaryLabel
          }) : null]
        })
      })]
    })
  );
};
BpkAutosuggestSuggestion.propTypes = {
  value: PropTypes.node.isRequired,
  subHeading: PropTypes.node,
  tertiaryLabel: PropTypes.string,
  icon: PropTypes.func,
  indent: PropTypes.bool,
  className: PropTypes.string
};
BpkAutosuggestSuggestion.defaultProps = {
  subHeading: null,
  tertiaryLabel: null,
  icon: null,
  indent: false,
  className: null
};
export default BpkAutosuggestSuggestion;
