// @skyscanner/backpack-web/bpk-component-barchart/src/BpkBarchartDefs.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */const GRADIENT_ATTRIBUTES = {
  x1: 0,
  y1: 0,
  x2: 0,
  y2: 1
};
const BpkBarchartDefs = () => /*#__PURE__*/_jsxs("defs", {
  children: [/*#__PURE__*/_jsxs("linearGradient", {
    id: "bpk-barchart__def-gradient",
    ...GRADIENT_ATTRIBUTES,
    children: [/*#__PURE__*/_jsx("stop", {
      offset: "0",
      stopColor: "white",
      stopOpacity: "0"
    }), /*#__PURE__*/_jsx("stop", {
      offset: "10%",
      stopColor: "white",
      stopOpacity: "1"
    })]
  }), /*#__PURE__*/_jsx("mask", {
    id: "bpk-barchart__def-mask",
    children: /*#__PURE__*/_jsx("rect", {
      x: "0",
      y: "0",
      width: "100%",
      height: "100%",
      fill: "url(#bpk-barchart__def-gradient)"
    })
  })]
});
export default BpkBarchartDefs;
