// @skyscanner/backpack-web/bpk-component-price/src/BpkPrice.js
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */import PropTypes from 'prop-types';
import BpkText, { TEXT_STYLES } from "../../bpk-component-text";
import { cssModules } from "../../bpk-react-utils";
import { SIZES, ALIGNS } from "./common-types";
import STYLES from "./BpkPrice.module.css";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
const getClassName = cssModules(STYLES);
const BpkPrice = props => {
  const {
    align,
    className,
    leadingClassName,
    leadingText,
    previousPrice,
    price,
    size,
    trailingText,
    ...rest
  } = props;
  const isSmall = size === SIZES.small;
  const isAlignRight = align === ALIGNS.right;
  return /*#__PURE__*/_jsxs("div", {
    className: getClassName('bpk-price', isAlignRight && 'bpk-price--right', className)
    // $FlowFixMe[cannot-spread-inexact] - inexact rest. See 'decisions/flowfixme.md'.
    ,
    ...rest,
    children: [/*#__PURE__*/_jsxs("div", {
      className: getClassName(previousPrice && 'bpk-price__leading', isAlignRight && 'bpk-price__leading--right', leadingClassName),
      children: [previousPrice && /*#__PURE__*/_jsx("span", {
        className: getClassName('bpk-price__previous-price'),
        children: /*#__PURE__*/_jsx(BpkText, {
          textStyle: isSmall ? TEXT_STYLES.xs : TEXT_STYLES.sm,
          tagName: "span",
          children: previousPrice
        })
      }), previousPrice && leadingText && /*#__PURE__*/_jsx("span", {
        className: getClassName('bpk-price__separator'),
        children: /*#__PURE__*/_jsx(BpkText, {
          textStyle: isSmall ? TEXT_STYLES.xs : TEXT_STYLES.sm,
          tagName: "span",
          children: "\uD802\uDD1F"
        })
      }), leadingText && /*#__PURE__*/_jsx(BpkText, {
        textStyle: isSmall ? TEXT_STYLES.xs : TEXT_STYLES.sm,
        tagName: "span",
        children: leadingText
      })]
    }), /*#__PURE__*/_jsxs("div", {
      className: getClassName(isAlignRight && 'bpk-price__column-container'),
      children: [/*#__PURE__*/_jsx("span", {
        className: getClassName('bpk-price__price', !isAlignRight && 'bpk-price__spacing'),
        children: /*#__PURE__*/_jsx(BpkText, {
          textStyle: isSmall ? TEXT_STYLES.heading4 : TEXT_STYLES.xxl,
          tagName: "span",
          children: price
        })
      }), trailingText && /*#__PURE__*/_jsx("span", {
        className: getClassName('bpk-price__trailing'),
        children: /*#__PURE__*/_jsx(BpkText, {
          textStyle: isSmall ? TEXT_STYLES.xs : TEXT_STYLES.sm,
          tagName: "span",
          children: trailingText
        })
      })]
    })]
  });
};
BpkPrice.propTypes = {
  price: PropTypes.string.isRequired,
  size: PropTypes.oneOf(Object.keys(SIZES)),
  align: PropTypes.oneOf(Object.keys(ALIGNS)),
  className: PropTypes.string,
  leadingText: PropTypes.string,
  trailingText: PropTypes.string,
  previousPrice: PropTypes.string,
  leadingClassName: PropTypes.string
};
BpkPrice.defaultProps = {
  size: SIZES.small,
  align: ALIGNS.left,
  className: null,
  leadingText: null,
  trailingText: null,
  previousPrice: null,
  leadingClassName: null
};
export default BpkPrice;
