// @skyscanner/backpack-web/bpk-component-nudger/src/BpkConfigurableNudger.js
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BpkButtonV2, BUTTON_TYPES } from "../../bpk-component-button";
import { withButtonAlignment } from "../../bpk-component-icon";
import MinusIcon from "../../bpk-component-icon/sm/minus";
import PlusIcon from "../../bpk-component-icon/sm/plus";
import { cssModules } from "../../bpk-react-utils";
import STYLES from "./BpkNudger.module.css";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
const getClassName = cssModules(STYLES);
const AlignedMinusIcon = withButtonAlignment(MinusIcon);
const AlignedPlusIcon = withButtonAlignment(PlusIcon);
const BpkConfigurableNudger = ({
  buttonType = 'secondary',
  className = null,
  compareValues,
  decreaseButtonLabel,
  decrementValue,
  formatValue,
  id,
  increaseButtonLabel,
  incrementValue,
  inputClassName = null,
  max,
  min,
  onChange,
  value,
  ...rest
}) => {
  const classNames = getClassName('bpk-nudger', className);
  const maxButtonDisabled = compareValues(value, max) >= 0;
  const minButtonDisabled = compareValues(value, min) <= 0;
  const inputStyles = getClassName('bpk-nudger__input', inputClassName && inputClassName, buttonType === 'secondaryOnDark' && 'bpk-nudger__input--secondary-on-dark');
  return /*#__PURE__*/_jsxs("div", {
    className: classNames,
    children: [/*#__PURE__*/_jsx(BpkButtonV2, {
      type: BUTTON_TYPES[buttonType],
      iconOnly: true,
      onClick: () => onChange(decrementValue(value)),
      disabled: minButtonDisabled,
      title: decreaseButtonLabel,
      "aria-controls": id,
      children: /*#__PURE__*/_jsx(AlignedMinusIcon, {})
    }), /*#__PURE__*/_jsx("input", {
      type: "text",
      "aria-live": "polite",
      readOnly: true,
      value: formatValue(value),
      id: id,
      className: inputStyles,
      ...rest
    }), /*#__PURE__*/_jsx(BpkButtonV2, {
      type: BUTTON_TYPES[buttonType],
      iconOnly: true,
      onClick: () => onChange(incrementValue(value)),
      disabled: maxButtonDisabled,
      title: increaseButtonLabel,
      "aria-controls": id,
      children: /*#__PURE__*/_jsx(AlignedPlusIcon, {})
    })]
  });
};
export default BpkConfigurableNudger;
