// @skyscanner/backpack-web/bpk-component-nudger/src/BpkNudger.js
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// @ts-expect-error Untyped import. See `decisions/imports-ts-suppressions.md`.
import BpkLabel from "../../bpk-component-label";
import BpkText, { TEXT_STYLES } from "../../bpk-component-text";
import { cssModules } from "../../bpk-react-utils";
import BpkConfigurableNudger from "./BpkConfigurableNudger";
import STYLES from "./BpkNudger.module.css";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
const getClassName = cssModules(STYLES);
const compareValues = (a, b) => a - b;
const incrementValue = currentValue => currentValue + 1;
const decrementValue = currentValue => currentValue - 1;
const formatValue = currentValue => currentValue.toString();
const BpkNudger = ({
  buttonType = 'secondary',
  className = null,
  icon,
  id,
  subtitle,
  title,
  ...rest
}) => {
  const classNames = getClassName(title && 'bpk-nudger__container');
  return /*#__PURE__*/_jsxs("div", {
    className: classNames,
    children: [title && /*#__PURE__*/_jsxs("div", {
      className: getClassName('bpk-nudger__label'),
      children: [/*#__PURE__*/_jsx(BpkLabel, {
        htmlFor: id,
        children: icon
      }), /*#__PURE__*/_jsxs("span", {
        // For a11y on IOS, role='text' forces label to be read in full. More info: https://axesslab.com/text-splitting/
        // eslint-disable-next-line jsx-a11y/aria-role
        role: "text",
        className: getClassName('bpk-nudger__label--title-subtitle'),
        children: [/*#__PURE__*/_jsx(BpkText, {
          textStyle: TEXT_STYLES.heading5,
          children: title
        }), subtitle && /*#__PURE__*/_jsx("span", {
          className: getClassName('bpk-nudger__label--subtitle'),
          children: /*#__PURE__*/_jsx(BpkText, {
            children: subtitle
          })
        })]
      })]
    }), /*#__PURE__*/_jsx(BpkConfigurableNudger, {
      inputClassName: getClassName('bpk-nudger__input--numeric'),
      compareValues: compareValues,
      incrementValue: incrementValue,
      decrementValue: decrementValue,
      formatValue: formatValue
      // TODO: className to be removed
      // eslint-disable-next-line @skyscanner/rules/forbid-component-props
      ,
      className: className,
      buttonType: buttonType,
      id: id,
      ...rest
    })]
  });
};
export default BpkNudger;
