// @skyscanner/backpack-web/bpk-component-navigation-bar/src/BpkNavigationBar.js
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import BpkText, { TEXT_STYLES } from "../../bpk-component-text";
import { cssModules } from "../../bpk-react-utils";
import STYLES from "./BpkNavigationBar.module.css";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
const getClassNames = cssModules(STYLES);
export const BAR_STYLES = {
  default: 'default',
  onDark: 'on-dark'
};
const BpkNavigationBar = props => {
  const {
    barStyle = BAR_STYLES.default,
    className,
    id,
    leadingButton,
    sticky = false,
    title,
    titleTagName = "span",
    titleTextStyle = TEXT_STYLES.heading5,
    trailingButton,
    ...rest
  } = props;

  // If the title is a component that sets its own id we want the aria-labelledby on the nav to match this so it can find the element
  // Otherwise if its just a string we set the id on the title component.
  const titleId = typeof title === 'string' ? `${id}-bpk-navigation-bar-title` : id;
  return /*#__PURE__*/_jsxs("nav", {
    "aria-labelledby": titleId,
    className: getClassNames('bpk-navigation-bar', `bpk-navigation-bar--${barStyle}`, sticky && 'bpk-navigation-bar__sticky', className),
    ...rest,
    children: [leadingButton && /*#__PURE__*/_jsx("div", {
      className: getClassNames('bpk-navigation-bar__leading-item', `bpk-navigation-bar__leading-item--${barStyle}`),
      children: leadingButton
    }), typeof title === 'string' ? /*#__PURE__*/_jsx("span", {
      className: getClassNames('bpk-navigation-bar__title', `bpk-navigation-bar__title--${barStyle}`),
      children: /*#__PURE__*/_jsx(BpkText, {
        id: titleId,
        textStyle: titleTextStyle,
        tagName: titleTagName,
        children: title
      })
    }) : /*#__PURE__*/_jsx("div", {
      className: getClassNames('bpk-navigation-bar__title-container'),
      children: title
    }), trailingButton && /*#__PURE__*/_jsx("div", {
      className: getClassNames('bpk-navigation-bar__trailing-item', `bpk-navigation-bar__trailing-item-${barStyle}`),
      children: trailingButton
    })]
  });
};
export default BpkNavigationBar;
