// @skyscanner/backpack-web/bpk-component-graphic-promotion/src/BpkGraphicPromo.js
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2022 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BpkButtonV2, BUTTON_TYPES } from "../../bpk-component-button";
import { cssModules } from "../../bpk-react-utils";
import STYLES from "./BpkGraphicPromo.module.css";
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
const getClassName = cssModules(STYLES);
const ACCESSIBILITY_KEYS = {
  Enter: 13 /* Enter */,
  Space: 32 /* Space */
};
const isAccessibilityClick = event => Object.keys(ACCESSIBILITY_KEYS).includes(event.key) || Object.values(ACCESSIBILITY_KEYS).includes(event.keyCode || event.which);
export const TEXT_ALIGN = {
  start: 'start',
  center: 'center',
  end: 'end'
};
const constructAriaLabel = ({
  buttonText,
  headline,
  sponsor,
  subheading,
  tagline
}) => {
  const text = [];
  const addText = value => value && text.push(value);
  if (sponsor) {
    addText(sponsor.label);
    addText(sponsor.altText);
  } else {
    addText(tagline);
  }
  addText(headline);
  addText(subheading);
  addText(buttonText);
  return text.join('. ');
};
const BpkGraphicPromo = ({
  buttonText,
  className = null,
  contentId,
  headline,
  invertVertically = false,
  onClick,
  sponsor = null,
  style = {},
  subheading = null,
  tagline = null,
  textAlign
}) => {
  // FIXME: Use useCallback() here when React is updated.
  const onClickWrapper = event => {
    event.stopPropagation();
    onClick();
  };
  const onKeyWrapper = event => isAccessibilityClick(event) && onClick();
  const cardClasses = getClassName('bpk-card', 'bpk-graphic-promo', className);
  const containerClasses = getClassName('bpk-graphic-promo__container', `bpk-graphic-promo__container--${textAlign}`, invertVertically && 'bpk-graphic-promo__container--inverted');
  const getTextClasses = baseClass => getClassName(baseClass, `${baseClass}--${textAlign}`);
  return (
    /*#__PURE__*/
    // The card appears as a single component for the screen reader; its children are hidden. The card handles mouse
    // clicks and key presses (Enter/Space) for the whole component, as described here:
    // https://developer.mozilla.org/en-US/docs/Web/Accessibility/Keyboard-navigable_JavaScript_widgets
    _jsx("div", {
      id: contentId || '',
      className: cardClasses,
      style: style,
      role: "link",
      "aria-label": constructAriaLabel({
        buttonText,
        headline,
        sponsor,
        subheading,
        tagline
      }),
      tabIndex: 0,
      onClick: onClickWrapper,
      onKeyDown: onKeyWrapper,
      children: /*#__PURE__*/_jsxs("div", {
        id: contentId && `${contentId}__content` || '',
        className: containerClasses,
        "aria-hidden": true,
        children: [/*#__PURE__*/_jsx("div", {
          className: getTextClasses('bpk-graphic-promo__sponsor-content'),
          children: sponsor && /*#__PURE__*/_jsxs(_Fragment, {
            children: [/*#__PURE__*/_jsx("span", {
              className: getClassName('bpk-graphic-promo__sponsor-label'),
              children: sponsor.label
            }), /*#__PURE__*/_jsx("img", {
              className: getClassName('bpk-graphic-promo__sponsor-logo'),
              alt: sponsor.altText,
              src: sponsor.logo
            })]
          })
        }), /*#__PURE__*/_jsxs("div", {
          className: getTextClasses('bpk-graphic-promo__promo-content'),
          children: [!sponsor && tagline && /*#__PURE__*/_jsx("span", {
            className: getClassName('bpk-graphic-promo__tagline'),
            children: tagline
          }), /*#__PURE__*/_jsx("h2", {
            className: getClassName('bpk-graphic-promo__headline'),
            children: headline
          }), subheading && /*#__PURE__*/_jsx("p", {
            className: getClassName('bpk-graphic-promo__subheading'),
            children: subheading
          }), /*#__PURE__*/_jsx(BpkButtonV2, {
            type: BUTTON_TYPES.primaryOnDark
            // TODO: className to be removed
            // eslint-disable-next-line @skyscanner/rules/forbid-component-props
            ,
            className: getClassName('bpk-graphic-promo__cta'),
            onClick: onClickWrapper,
            tabIndex: -1 /* button is not focusable for accessibility */,
            children: buttonText
          })]
        })]
      })
    })
  );
};
export default BpkGraphicPromo;
